plonky2/src/hash/keccak.rs

```diff
@@ -1,50 +1,58 @@
 use std::iter;
 use std::mem::size_of;
 
-use itertools::Itertools;
-use keccak_hash::keccak;
+use tiny_keccak::{Hasher as _, Keccak};
 
 use crate::hash::hash_types::{BytesHash, RichField};
 use crate::hash::hashing::{PlonkyPermutation, SPONGE_WIDTH};
 use crate::plonk::config::Hasher;
-use crate::util::serialization::Buffer;
 
 /// Keccak-256 pseudo-permutation (not necessarily one-to-one) used in the challenger.
 /// A state `input: [F; 12]` is sent to the field representation of `H(input) || H(H(input)) || H(H(H(input)))`
 /// where `H` is the Keccak-256 hash.
 pub struct KeccakPermutation;
 impl<F: RichField> PlonkyPermutation<F> for KeccakPermutation {
     fn permute(input: [F; SPONGE_WIDTH]) -> [F; SPONGE_WIDTH] {
-        let mut state = vec![0u8; SPONGE_WIDTH * size_of::<u64>()];
-        for i in 0..SPONGE_WIDTH {
-            state[i * size_of::<u64>()..(i + 1) * size_of::<u64>()]
-                .copy_from_slice(&input[i].to_canonical_u64().to_le_bytes());
+        // Absorb input
+        let mut sponge = Keccak::v256();
+        for input in input {
+            sponge.update(&input.to_canonical_u64().to_le_bytes());
         }
 
-        let hash_onion = iter::repeat_with(|| {
-            let output = keccak(state.clone()).to_fixed_bytes();
-            state = output.to_vec();
-            output
+        // Create output iterator by iterating hash function
+        let mut state = [0_u8; 32];
+        sponge.finalize(&mut state);
+        let states = iter::successors(Some(state), |state| {
+            let mut next = [0_u8; 32];
+            let mut sponge = Keccak::v256();
+            sponge.update(state);
+            sponge.finalize(&mut next);
+            Some(next)
         });
 
-        let hash_onion_u64s = hash_onion.flat_map(|output| {
-            output
-                .chunks_exact(size_of::<u64>())
-                .map(|word| u64::from_le_bytes(word.try_into().unwrap()))
-                .collect_vec()
-        });
-
-        // Parse field elements from u64 stream, using rejection sampling such that words that don't
-        // fit in F are ignored.
-        let hash_onion_elems = hash_onion_u64s
-            .filter(|&word| word < F::ORDER)
-            .map(F::from_canonical_u64);
-
-        hash_onion_elems
-            .take(SPONGE_WIDTH)
-            .collect_vec()
-            .try_into()
-            .unwrap()
+        // Collect SPONGE_WIDTH elements using rejection sampling
+        // Note: ideally we'd do this elegantly with a `flat_map`, but that
+        // is challenging to make allocation free.
+        let mut result = [F::ZERO; SPONGE_WIDTH];
+        let mut result_iter = result.iter_mut();
+        let mut next = result_iter.next();
+        for state in states {
+            if next.is_none() {
+                break;
+            }
+            for chunk in state.chunks(size_of::<F>()) {
```

Comment:
I think it'd be best to use `size_of::<u64>` to ensure the conversion to u64 below succeeds. Not an issue with our current field, but just in case `size_of::<F>() < 8` for some future field